src/devices/cpu/drcbearm64.cpp

```diff
@@ -0,0 +1,3876 @@
+	const parameter &sizep = inst.param(2);
+	assert(sizep.is_size());
+	const auto size = sizep.size();
+
+	const a64::Gp dstreg = dstp.select_register(TEMP_REG2, inst.size());
+
+	if ((1 << size) >= inst.size())
+	{
+		mov_param_param(a, inst.size(), dstp, srcp);
+	}
+	else
+	{
+		if (srcp.is_memory())
+		{
+			if (size == SIZE_BYTE)
+				emit_ldrsb_mem(a, dstreg.x(), srcp.memory());
+			else if (size == SIZE_WORD)
+				emit_ldrsh_mem(a, dstreg.x(), srcp.memory());
+			else if (size == SIZE_DWORD)
+				emit_ldrsw_mem(a, dstreg.x(), srcp.memory());
+			else if (size == SIZE_QWORD)
+				emit_ldr_mem(a, dstreg.x(), srcp.memory());
+		}
+		else
+		{
+			const a64::Gp tempreg = srcp.select_register(dstreg, 8);
+			mov_reg_param(a, inst.size(), tempreg, srcp);
+
+			if (size == SIZE_BYTE)
+				a.sxtb(dstreg.x(), tempreg.w());
+			else if (size == SIZE_WORD)
+				a.sxth(dstreg.x(), tempreg.w());
+			else if (size == SIZE_DWORD)
+				a.sxtw(dstreg.x(), tempreg.w());
+		}
+
+		mov_param_reg(a, inst.size(), dstp, dstreg);
+	}
+
+	if (inst.flags())
+		a.tst(dstreg, dstreg);
+}
+
+void drcbe_arm64::op_roland(a64::Assembler &a, const uml::instruction &inst)
+{
+	assert(inst.size() == 4 || inst.size() == 8);
+	assert_no_condition(inst);
+	assert_flags(inst, FLAG_S | FLAG_Z);
+
+	be_parameter dstp(*this, inst.param(0), PTYPE_MR);
+	be_parameter srcp(*this, inst.param(1), PTYPE_MRI);
+	be_parameter shiftp(*this, inst.param(2), PTYPE_MRI);
+	be_parameter maskp(*this, inst.param(3), PTYPE_MRI);
+
+	if (!maskp.is_immediate() || (maskp.is_immediate() && !maskp.is_immediate_value(0)))
+	{
+		const a64::Gp scratch = select_register(TEMP_REG1, inst.size());
+
+		mov_reg_param(a, inst.size(), scratch, srcp);
+
+		if (shiftp.is_immediate())
+		{
+			const auto shift = ((inst.size() * 8) - (shiftp.immediate() % (inst.size() * 8))) % (inst.size() * 8);
+
+			if (shift != 0)
+				a.ror(scratch, scratch, shift);
+		}
+		else
+		{
+			const a64::Gp scratch2 = select_register(SCRATCH_REG1, inst.size());
+			const a64::Gp shift = shiftp.select_register(TEMP_REG2, inst.size());
+
+			mov_reg_param(a, inst.size(), shift, shiftp);
+
+			a.mov(scratch2, inst.size() * 8);
+			a.sub(shift, scratch2, shift);
+			a.ror(scratch, scratch, shift);
+		}
+
+		const a64::Gp output = dstp.select_register(TEMP_REG1, inst.size());
+		if (maskp.is_immediate() && is_valid_immediate_mask(maskp.immediate(), inst.size()))
+		{
+			a.ands(output, scratch, maskp.immediate());
+		}
+		else
+		{
+			const a64::Gp mask = maskp.select_register(TEMP_REG2, inst.size());
+			mov_reg_param(a, inst.size(), mask, maskp);
+
+			a.ands(output, scratch, mask);
+		}
+
+		mov_param_reg(a, inst.size(), dstp, output);
+	}
+	else
+	{
+		const a64::Gp zero = select_register(a64::xzr, inst.size());
+
+		mov_param_reg(a, inst.size(), dstp, zero);
+
+		if (inst.flags())
+			a.tst(zero, zero);
+	}
+}
+
+void drcbe_arm64::op_rolins(a64::Assembler &a, const uml::instruction &inst)
+{
```

Comment:
Is it worth detecting cases where this can be special-cased to use a `bfi` instruction?  If the mask and shift are immediate, `bfi` can be used if the mask is contiguous and starts at the bit position equal to the shift value.

For example:
* `ROLINS(d, s, 7, 0x0380)` is equivalent to `bfi(d, s, 7, 3)`

In cases where the shift and mask are immediate, the mask is contiguous, and the mask does not start at the bit position equal to the shift value, you can use `ror` followed by `bfi`.

For example, `ROLINS(d, s, 4, 0x1f80)` is equivalent to `ror(tmp, s, 3)` followed by `bfi(d, tmp, 7, 6)`

Reply:
>Is it worth detecting cases where this can be special-cased to use a bfi instruction? If the mask and shift are immediate, bfi can be used if the mask is contiguous and starts at the bit position equal to the shift value.

Two things to note:
- If the rolins function is called with all immediates and does not use flags, it gets optimized out. This applies to most opcodes if they use constants https://github.com/mamedev/mame/blob/master/src/devices/cpu/uml.cpp#L436-L474
- My current implementation will simplify shift + mask constants down to ror or mov as appropriate already

But the use of `bfi` to simplify the second half of the code probably will work so I'll try it.

Reply:
My implementation of the bfi thing broke rolins. bfi requires the value to start at bit 0 to insert it into the destination register, so i just seems like too much effort to save 1 instruction so I'm going to revert that change.